### PR TITLE
Extend and clean up skill DB code

### PR DESCRIPTION
**Pull request by Kenpachi2k13**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

- Add defaults, bounds and validations when reading skill DB.
- Change maximum length of skill descriptions to `50`.
- Change maximum skill level to `20`.
- Remove `FixedCastTime` settings from pre-RE skill DB.
- Cap `SkillInstances` to `25` (`MAX_SKILLUNITGROUP`) in (pre-)RE skill DB.
- Remove unused variable `nocast` from `struct s_skill_db`.
- Make `Hit`, `AttackType`, `InterruptCast`, `CastDefRate`, `Requirements->State`, `Requirements->Items->Amount`, `Unit->Id`, `Unit->Interval` and `Unit->Target` can be grouped by levels.
- Add missing types to `Unit->Target`.
- Remove hard-coded required item amounts and update skill DBs accordingly.
- Implement `Requirements->Items->Any` code.
- Implement `Requirements->Equip` code.
- Remove hard-coded required equipment and update skill DBs accordingly.
- Fix `Requirements->MaxHPTrigger` code in `skill_check_condition_mercenary()`.
- Implement `Requirements->MaxSPTrigger` code.
- Update entry structure in skill DBs.

**Issues addressed:** #1250

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/skill.c [hunk 1]
```diff
@@ -20884,140 +21929,456 @@ static int skill_validate_ammotype_sub(const char *type, bool on, struct s_skill
...
 	if ((tt = libconfig->setting_get_member(conf, "AmmoTypes")) && config_setting_is_group(tt)) {
 		int j = 0;
 		struct config_setting_t *amt = { 0 };
 		while ((amt = libconfig->setting_get_elem(tt, j++))) {
 			if (skill->validate_ammotype_sub(config_setting_name(amt), libconfig->setting_get_bool_real(amt), sk))
-				skilldb_invalid_error(config_setting_name(amt), config_setting_name(tt), sk->nameid);
+				ShowWarning("%s: Invalid required ammunition type %s specified for skill ID %d in %s! Skipping type...\n",
+					    __func__, config_setting_name(amt), sk->nameid, conf->file);
+		}
+	} else if( libconfig->setting_lookup_string(conf, "AmmoTypes", &tstr)) {
```
[the remark this answers is not in the document]
Comment (Kenpachi2k13) [reply src/map/skill.c]: Was fixed in following commit. (https://github.com/HerculesWS/Hercules/pull/2731/commits/eca478b48c884d7399c7eecc8856d735f0abb83d)

File: src/map/skill.c [hunk 2]
```diff
@@ -21126,61 +22487,108 @@ static int skill_validate_unit_flag_sub(const char *type, bool on, struct s_skil
...
+	nullpo_retv(conf);
+	nullpo_retv(sk);
+
+	sk->unit_target = BCT_NOONE;
+
+	const char *type = NULL;
+
+	if(libconfig->setting_lookup_string(conf, "Target", &type)) {
+
+		if(!strcmpi(type,"NotEnemy")) sk->unit_target = BCT_NOENEMY;
```
[the remark this answers is not in the document]
Comment (Kenpachi2k13) [reply src/map/skill.c]: Was fixed in following commit. (https://github.com/HerculesWS/Hercules/pull/2731/commits/eca478b48c884d7399c7eecc8856d735f0abb83d)

File: src/map/skill.c [hunk 3]
```diff
@@ -15619,23 +15619,11 @@ static struct skill_condition skill_get_requirement(struct map_session_data *sd,
 	}
 
 	for( i = 0; i < MAX_SKILL_ITEM_REQUIRE; i++ ) {
-		int item_idx = (skill_lv - 1) % MAX_SKILL_ITEM_REQUIRE;
```
[the remark this answers is not in the document]
Comment (Kenpachi2k13) [reply src/map/skill.c:15489]: I guess you're talking about https://github.com/HerculesWS/Hercules/pull/2731/commits/fe83b90763c6e40bf0aeab565a8b975181037e07 and https://github.com/HerculesWS/Hercules/pull/2731/commits/998b585f87c88d1629058e2d518518dffc9f36c2.

I included the DB changes on purpose to be able to compare the conditions which were removed from the code with the conditions I added to the DB files.

**Comment by Kenpachi2k13** [comment]
> also because was big changed in skill db,

Actually the changes are not that big. It just gives the impression because I outsourced many things into new functions and because I applied code style to existing functions.

> probably need add skill db dumper before this pr and after this pr (excluding db files changes). and compare output. output must be same

This won't work because some skills or parts of a skill won't be loaded due to the validations I added to the code and thus the dumps will never be identical.

**Comment by SombraRO** [comment]
@Kenpachi2k13 wouldn't it be nice to add the use of `struct block_list *bl` in the `skill_get` functions?

would give more options for changes using the plugins, for example I can set the **Fire Bolt** ability to give **20 hits** when used by a **poring**

**Comment by Kenpachi2k13** [comment]
All the data used by the `skill_get*` functions is stored in the skill itself. You can't set/read it for a specific monster/character.

**Comment by SombraRO** [comment]
@Kenpachi2k13 It would work because it would make a reading that could be another db or a defined value.

Every time a skill is executed, the skill_get function is executed, so you can check and set a value

```
if(bl->type == BL_MOB)
	return 20;
else
	return skill->dbs->db[idx].hit;
```

**Comment by Kenpachi2k13** [comment]
Oh. Now I get what you mean. Sorry. :sweat_smile:
Yes, this is possible of course. I'll talk to @MishimaHaruna about implementing this.

**Pull request merged**